WIP: Initial work on adding packml_stacklight

winchesterag opened the pull request:
- [x] code / design review
- [x] add config/startup parameters and node testing
- [x] test subscribe and publish
- [x] test actual node with hardware

Commit:
feat: wip added initial structure and classes for packml_stacklight

Commit:
feat: added unit tested and more util functions to keep work seperate from ros node

Commit:
feat: continued stacklight work, added std_msg dependency and added calls to utils from ros node

Commit:
feat: removed unused process_sates, added test for publish topics

shaun-edwards commented on packml_stacklight/include/packml_stacklight/utils.h:
```diff
@@ -0,0 +1,103 @@
+  BLUE = 4,
+};
+
+static std::map<LightValues::LightValue, std::string> LightValueMap = { { UNDEFINED, "UNDEFINED-LIGHT" },
```
Variable naming doesn't meet style guide.

shaun-edwards commented on packml_stacklight/src/packml_stacklight.cpp:
```diff
@@ -0,0 +1,83 @@
+{
+PackmlStacklight::PackmlStacklight(ros::NodeHandle nh, ros::NodeHandle pn) : nh_(nh), pn_(pn)
+{
+  ros::NodeHandle packml_node("~/packml");
```
This is a bit odd....you don't normally need a reference to another node.

shaun-edwards commented on packml_stacklight/src/packml_stacklight.cpp:
```diff
@@ -0,0 +1,83 @@
+{
+  ros::NodeHandle packml_node("~/packml");
+
+  status_sub_ = packml_node.subscribe<packml_msgs::Status>("status", 1, &PackmlStacklight::callBackStatus, this,
```
Any subscribers should use the passed in `nh`

shaun-edwards replied:
`ros::TransportHints().reliable().tcpNoDelay(true)` should only be used on high frequency topics.

shaun-edwards commented on packml_stacklight/src/packml_stacklight.cpp:
```diff
@@ -0,0 +1,83 @@
+
+  packml_msgs::State temp;
+  temp.val = packml_msgs::State::UNDEFINED;
+  StatusAction* status_ptr = getActionFromState(temp);
```
Does this need to be a pointer?  We should only use pointers where required?

Commit:
feat: added startup and config params, handled polling frequency

Commit:
feat: moved most of utils stuff to class with exception of existing structs and enums

updated test to Test fixtures along with utils class

Commit:
refactor: converted structs to classes and put each container class in own header file

Commit:
feat: added check for no status timeout to reset to unspecified

fix: flash check when never receive status

test: added test to test specific case where bug was found, added test for new status timeout

shaun-edwards commented on packml_stacklight/CMakeLists.txt:
```diff
@@ -0,0 +1,75 @@
+	target_link_libraries(${PROJECT_NAME}_test ${PROJECT_NAME} ${catkin_LIBRARIES})
+endif()
+
+#if(CATKIN_ENABLE_TESTING)
```
Please delete commented items

shaun-edwards commented on packml_stacklight/CMakeLists.txt:
```diff
@@ -0,0 +1,75 @@
+#  target_link_libraries(${PROJECT_NAME}_utest ${PROJECT_NAME} ${catkin_LIBRARIES})
+#endif()
+
+#install(TARGETS ${PROJECT_NAME}
```
Install commands are required

winchesterag replied:
I have added in the install commands in, but not sure how to test.
I also am not familiar with what is needed, however used what other packages had.

shaun-edwards reviewed (changes requested):
@winchesterag, just a couple of items to clean up.  Overall, looks good!

Commit:
ci:removed commented items, added install lines, updated author

Commit:
ci: tweaked CMakeLists.txt and commented out orig

Commit:
wip:testing build issues that require ros::Time

Commit:
wip: commented out ros::init and added in ros::Time::init

Commit:
wip: added back in commented out test

Commit:
ci: removed commented out test section, removed test launch file

130s commented:
CI failing. E.g. https://travis-ci.org/ros-industrial-consortium/packml/jobs/539897337#L2317

```
/root/catkin_ws/devel/.private/packml_stacklight/lib/packml_stacklight/packml_stacklight_utest --gtest_output=xml:/root/catkin_ws/build/packml_stacklight/test_results/packml_stacklight/gtest-packml_stacklight_utest.xml
[FATAL] [1559338761.464739287]: ROS_MASTER_URI is not defined in the environment. Either type the following or (preferrably) add this to your ~/.bashrc file in order set up your local machine as a ROS master:
export ROS_MASTER_URI=http://localhost:11311
then, type 'roscore' in another shell to actually launch the master program.
Segmentation fault (core dumped)
:
build/packml_stacklight/test_results/packml_stacklight/MISSING-gtest-packml_stacklight_utest.xml: 1 tests, 0 errors, 1 failures, 0 skipped
Summary: 6 tests, 0 errors, 1 failures, 0 skipped
/root/ici/tests/source_tests.sh: line 244: error: command not found
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
Function 'catkin_run_tests' returned with code '127' after 0 min 19 sec
```
Direct cause is I think a test case code looks for ROS Master but `ROS_MASTER_URI` environment variable, which tells your ROS programs the location of the Master, is not found. That happened probably because there's `ros::init` at [packml_stacklight/test/test_stacklight.cpp](https://github.com/ros-industrial-consortium/packml/blob/8cf05d33f7ce4eb2b47599883fb0b6d87ae2d099/packml_stacklight/test/test_stacklight.cpp#L768), which I think tries to connect to ROS Master. There seems to be a launch file for testing ([test_stacklight.launch](https://github.com/ros-industrial-consortium/packml/blob/ef489bea7eaeab9c1dedd3bb73852e469a6d14f5/packml_stacklight/test/launch/test_stacklight.launch)), which starts ROS Master, but in your [testing setting the launch file isn't used](https://github.com/ros-industrial-consortium/packml/blob/ef489bea7eaeab9c1dedd3bb73852e469a6d14f5/packml_stacklight/CMakeLists.txt#L63). If you run the launch file, at least you'll see something differently as ROS Master should start.

Configuring `rostest` and reading its log is tricky, but [this on docs.ros.org](http://docs.ros.org/kinetic/api/catkin/html/howto/format2/rostest_configuration.html) is the resource I always go back to.
